README.md
haggis
======

This is a library of general purpose utility functions and classes written in
Python. Much of this code is intended to support the imprint project
<https://github.com/madphysicist/imprint>. The modules of this library are
more-or-less standalone utilities, organized by category.

This library contains plenty of code that relies on external programs and less
common Pyhon libraries. The code is considered to be "extras", in the setuptools
sense. Extras can be omitted without any modification to the python code. The
following extras are supported:

  - [docx]: Support for docx file utilities requires python-docx to be
    installed.
  - [latex]: Supporting LaTeX requires a host of external programs to work
    properly, so this feature is optional. See the dependency page in the main
    documentation for more information.
  - [pdf]: Requires the poppler library to be installed.
  - [ps]: Requires GhostScript to be installed.
  - [plot]: Plotting tools require matplotlib, which is a heavy dependency, and
    unnecessary for many purposes.
  - [term]: Terminal colors work out of the box on Linux, but require the
    colorama library to work on Windows.
  - [scio]: Science I/O depends on the scipy and astropy libraries to deliver
    IDL and FITS file support, respectively.
  - [xlsx]: Support for xlsx file utilities requires openpyxl to be installed.

To install extras such as latex and pdf, do::

    pip install haggis[latex,pdf]

See the documentation at <https://haggis.readthedocs.io/en/latest> for more
information, including the API documentation.

This library in licensed under the AGPLv3, and compatible with later versions.
You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.


CHANGELOG
Changelog
=========

0.10.0 (?)
----------
...

0.9.0 (2023-03-15)
------------------

* New functions and classes:
  - logs.add_detail_level
* Bugfixes:
  - Added missing default kwargs to monkeypatched version of
    logging.LogAdapter method in logs.add_log_level
  - Updated bondary handling in npy_util.prune_mask
* logs.add_logging_level verifies exception info settings if prior calls
  are found

0.8.0 (2022-11-04)
------------------

* New functions and classes:
  - recipes.chained_getter
* New unit tests:
  - tests.test_logs.TestAddLoggingLevel
  - tests.test_logs.TestResetHandlers
  - tests.test_logs.TestLogMaxFilter
  - tests.test_logs.TestMetaLoggableType
  - tests.test_recipes.TestChainedGetter
* Bugfixes:
  - Fixed sense of inclusive in logs.LogMaxFilter
* logs.add_logging_level updates logging.LoggerAdapter


0.7.1 (2022-10-24)
------------------

* New functions and classes:
  - npy_util.prune_mask


0.7.0 (2022-10-17)
------------------

* New functions and classes:
  - npy_util.masked_index
  - npy_util.unmasked_index
  - npy_util.find_peaks
* New unit tests:
  - tests.test_npy_util.TestMaskedIndex
  - tests.test_npy_util.TestUnmaskedIndex
  - tests.test_npy_util.TestFindPeaks


0.6.0 (2022-10-11)
------------------

* API changes:
  - Parameter stop in recipes.RangeBuilder.update is now optional
  - Property span in recipes.RangeBuilder
* Bugfixes:
  - Fixed handling of bulleted list styles in files.docx.list_number


0.5.0 (2022-06-17)
------------------

* New functions and classes:
  - multiline_repr in string_util
  - RangeBuilder in recipes


0.4.0 (2022-06-02)
------------------

* New functions and classes:
  - npy_util.replace_field
* New unit tests:
  - tests.test_npy_util.TestReplaceField


0.3.0 (2022-03-29)
------------------

* New functions and classes:
  - load.load_module no longer nested in load.module_as_dict
  - npy_util.iterate_dtype
* Bugfixes:
  - files.PreOpenedFile object is now properly iterable
* New unit tests:
  - tests.test_load.TestLoadModule
  - tests.test_load.TestModuleAsDict


0.2.0 (2022-01-27)
------------------

* New modules:
  - npy_util
  - tests.test_numbers
  - tests.test_objects
* New functions and classes:
  - getsizeof, size_type_mapping in objects
  - ndarray_handler added to size_type_mapping in objects
  - isolate_dtype in npy_util
  - as_base, digit_count in numbers
* New unit tests:
  - tests.test_objects.TestGetsizeof
  - tests.test_numbers.TestAsBase
  - tests.test_numbers.TestDigitCount
* Added docs to sdist
* Moved map_array, mask2runs, runs2mask from math to npy_util


0.1.1 (2022-01-16)
------------------

Minor updates to first release.

* New functions and classes:
  - segment_distance in math
* Bugfixes:
  - math.read_divide correctly handles scalar inputs
* Created pytest configuration, added first set of unit tests
  - files.tests.test_csv
  - tests.test_math.TestSegmentDistance
* Improved documentation with autosummary
* Added this CHANGELOG to installed data


0.1 (2022-01-10)
----------------

First release.

* New modules
  - time
  - files.ui
* New functions and classes:
  - json_registry made public in configuration
  - semilogx_base, semilogy_base, loglog_base in mpl_util
  - map_array, rms in math
  - split_extension in files
* API changes:
  - string_util.timestamp moved to time
  - Added mask2runs and runs2mask to math
  - Improvements to NumpyObject in configuration
  - Improvements to copy_func in objects
  - Added n argument to insert_suffix in files
* Bufixes:
  - mpl_util.show_extents axis confusion
  - Broken imports in submodules
* Officially supporting Python versions up to 3.10
* README -> README.md so GitHub can pick it up


0.1a2 (2021-01-09)
------------------

Second alpha release.

* Major additions to API
* New modules:
  - threads
  - configuration
  - structures
  - mapping
  - ctypes_util
* New functions and classes:
  - CloseableMixin in recipes
  - to_hex, camel2snake, snake2camel, timestamp in string_util
  - set_labels in mpl_utils
  - real_divide, threshold, ang_diff_* in math
* API changes:
  - Moved half of recipes to mapping
  - Updated Namespace, (in mapping), added setdefaults


0.1a1 (2021-04-03)
------------------

First alpha release.

* API changes:
  - SentinelType made public in root package
  - ensure_extension.modes renamed to ensure_exntension.policies in files


0.1a1.dev2 (2019-03-31)
-----------------------

Second development pre-release.

*   Major overhaul of documentation

0.1a1.dev1 (2019-03-20)
-----------------------

First development pre-release.

* First release on PyPI.
* Created Read the Docs pages.
* Only code required for imprint present initially.
* Documentation 90% complete.
* No tests worth speaking of.


src/haggis/version.py
# -*- coding: utf-8 -*-

# haggis: a library of general purpose utilities
#
# Copyright (C) 2019  Joseph R. Fox-Rabinovitz <jfoxrabinovitz at gmail dot com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# Author: Joseph Fox-Rabinovitz <jfoxrabinovitz at gmail dot com>
# Version: 13 Apr 2019: Initial Coding
# Version: 09 Jan 2021: Bumped to 0.1a2
# Version: 10 Jan 2022: Bumped to 0.1
# Version: 16 Jan 2022: Bumped to 0.1.1
# Version: 16 Jan 2022: Bumped to 0.2.0rc1: Documented versioning a bit better
# Version: 27 Jan 2022: Bumped to 0.2.0
# Version: 27 Jan 2022: Bumped to 0.3.0rc1
# Version: 29 Jan 2022: Decided to open backport 0.2.1 branch after all.
# Version: 29 Mar 2022: Bumped to 0.3.0
# Version: 29 Mar 2022: Bumped to 0.4.0rc1
# Version: 29 Mar 2022: Opened to 0.3.1 backport branch
# Version: 02 Jun 2022: Bumped to 0.4.0
# Version: 02 Jun 2022: Bumped to 0.5.0rc1
# Version: 17 Jun 2022: Bumped to 0.5.0
# Version: 17 Jun 2022: Bumped to 0.6.0rc1
# Version: 11 Oct 2022: Bumped to 0.6.0
# Version: 11 Oct 2022: Bumped to 0.7.0rc1
# Version: 17 Oct 2022: Bumped to 0.7.0
# Version: 17 Oct 2022: Bumped to 0.8.0rc1
# Version: 24 Oct 2022: Bumped to 0.7.1
# Version: 24 Oct 2022: Bumped back to 0.8.0rc1
# Version: 04 Nov 2022: Bumped to 0.8.0, part of proper release branch scheme
# Version: 04 Nov 2022: Bumped to 0.9.0rc1
# Version: 15 Mar 2023: Bumped to 0.9.0
# Version: 15 Mar 2023: Bumped to 0.10.0rc1


"""
:py:mod:`haggis.version` holds the version of the package.

The actual version number is normally exported through the top-level
:py:mod:`haggis` package, but defined here. This module exists to
facilitate deployment and testing, so users do not need to install the
entire package to get the version number.

Bump the version in the dev branch every time a prior version is tagged
in master. Unless a major rewrite is planned, the new version should
increment the minor number. The new version should be a release
candidate, and have the corresponding `rc1` suffix. All following
commits belong to the new version. Bump to full version immediately
before merging with master. A tag is assigned on the master commit that
merges in dev. Bump the version in dev immediately after merging master
back in.
"""


#: The current version
__version__ = '0.10.0rc1'
